### PR TITLE
Demonstrate the usage of extra mesh integers

### LINKED ISSUE #13764: Element extra integer
## Reason
We have pushed changes in libMesh for supporting the extra integers for elements (https://github.com/libMesh/libmesh/issues/2095). Now we need to enable this in MOOSE.
These capability can help radiation transport where complicated material assignment can happen and using subdomains for such assignment is not a valid option.

## Design
I am going to add a parameter with a list of integer names that we are going to add in `libMesh::Elem`. I will add few objects and modify one mesh modifier to demonstrate the usage of the extra integers. A test will be added in MOOSE as well. Applications like Rattlesnake, MAMMOTH are going to use this capability and have tests, but it is good to let MOOSE have tests as well.

## Impact
New capability.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #13764.

libMesh has been patched by @roystgnr .

The original contents in this PR is:

I finally got time to try using the mesh extra integers we added not long ago in libMesh. But I got a segfault in opt mode with the input in this PR. In debug mode, I hit an assertion in `dof_object.C:191`
```
libmesh_assert_equal_to(nei, this->n_extra_integers());
```
I must be misunderstanding something or there is some thing not very right. Need help.

### CODE REVIEW

**Comment by roystgnr** [comment]
Yikes!  I'm not sure how so simple a case could have passed unit testing but failed in practice, yet at a quick first glance your code looks fine and that assertion looks like an internal error in my code.

Let me see if I can replicate this.  I'm still on vacation (and on the road) tomorrow, and I'll probably be catching up on other urgent things on Thursday, but ping me again Thursday afternoon if I haven't gotten back to you by then?

**Comment by YaqiWang** [comment]
Sure. Thanks @roystgnr .

**Comment by YaqiWang** [comment]
@roystgnr any progress on this? Just checking.

**Comment by roystgnr** [comment]
Found time to replicate it, but no progress understanding it.

At the point the exception throws, the elements in the mesh here don't have extra integers assigned.  But that doesn't make sense to me, because the mesh itself definitely has an extra integer id added, and it was definitely added before any elements were created (although in hindsight I realize that we'd be more user-friendly if we didn't require that or if we at least screamed when someone violated the requirement), and I can set a breakpoint and *watch* the extra integers get added and then verify that they were correctly added.

So there's got to be some place where the extra integers are getting removed somehow; I'm still hunting for where and I still can't imagine how.  The first possibilities that come to mind are use cases that I specifically added unit tests for.

**Comment by roystgnr** [comment]
Actually, I take that back.  There is an important API not in unit testing and I think it's the code path MOOSE is using.  Let me try adding a unit test for it and see if that replicates the failure with libMesh alone.

**Comment by YaqiWang** [comment]
This PR is ready for review. It requires a libMesh update.

**Comment by YaqiWang** [comment]
@permcody This PR is updated and should be ready for review. Thanks.

**Review by permcody — COMMENTED** [review]
We need to have a meeting about this before we turn it loose on everyone. I want to make sure that this is the way we want to do this from a framework level.

@idaholab/moose-team - Let's make sure we chat about this next week!

**Comment by YaqiWang** [comment]
The only thing I really need is `MooseMesh:: addElementInteger` function. But I think having the regressions in framework is a good idea. I was hoping this would be a quick chat...

**Review by friedmud — CHANGES_REQUESTED** [review]
This is a really good start - but it's not quite there yet.  A few things are needed:

1.  The word "integer" shouldn't be used anywhere: These are "Extra IDs"
2.  There needs be an "ExtraIDInterface" that all objects inherit from to be able to get IDs (so no user code has to call through multiple layers of libMesh directly).  Also: it should handle our input parameters naturally: you should call "getExtraID("parameter_name") and it should do the right thing.
3.  More corner case handling... like checking for duplicate names, etc.

Like I say: this is good - just needs some more design work done.

Thanks!

Derek

**Comment by YaqiWang** [comment]
Dose calling `_current_elem->get_extra_integer(_id)` look ok to you in objects? I feel `ExtraIDInterface` is a little overkill. But I am ok to add it. It will not contain much codes though.

**Comment by friedmud** [comment]
The `ExtraIDInterface` is for getting/setting the `_id`.  It would replace this: 

```
_mesh.getMesh().get_elem_integer_index(getParam<std::string>("id_name")
```

With:

```
getExtraID("id_name")
```

Also: `MeshModifiers` are completely frozen at this point.  Just update the Generators.

**Comment by YaqiWang** [comment]
This is ready for review again. Thanks.

**Comment by permcody** [comment]
You are missing a few headers in your new class. Look at the unity build and see if you can fix them all. You can always build non-unity locally if you want.

**Comment by YaqiWang** [comment]
I think this is ready again. @friedmud

**Review by permcody — CHANGES_REQUESTED** [review]
This isn't quite the design @friedmud was after. He was proposing that the interface encapsulate the underlying API. What this means is that each of the places where we might access this underlying API will need to inherit from that new interface, then we'd just call `getElementID(id)`. If we succeed with the design, the end usage of the API will be very simple and straightforward without a lot of branching, dealing with string manipulation, etc at the end use point. The interface should handle all of that for you. I feel like you still have a little ways to go before this is complete. If you need help with this, let me know. Perhaps we can help you finish it.

**Comment by permcody** [comment]
@YaqiWang - I agree with your comment. Let's get this in. I'll review it once you get it rebased.

Thanks!

**Review by friedmud — CHANGES_REQUESTED** [review]
Hmmm - I like what you've done here.  Can we go one step further?

I would like to do away with the `_current_elem->get_extra_integer_id(_id)` call... ideally `_id` would already BE the extra integer ID.

However, that means we will need this to be "late binding".  `getElementID` will register with `MooseMesh` the element IDs that need to updated on each element and return a reference.  Druing `reinit(Elem)` the mesh will update each of the references for the current element.

**Comment by YaqiWang** [comment]
Working on this. Should update this PR very soon.

**Comment by YaqiWang** [comment]
Ready for new review.

File: framework/src/base/Assembly.C [hunk 1]
```diff
@@ -665,6 +667,9 @@ Assembly::reinitFE(const Elem * elem)
     }
   }
 
+  for (unsigned int i = 0; i < _extra_elem_ids.size(); ++i)
```
Comment (YaqiWang) [inline framework/src/base/Assembly.C]: have to make a copy here because libMesh currently does not provide an interface function to return the reference of these extra integers partly because they are within a larger array.

File: framework/src/interfaces/ElementIDInterface.C [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  unsigned int id = mesh_base->get_elem_integer_index(p[comp]);
+
+  return id;
+}
+
+const unsigned int &
+ElementIDInterface::getElementID(const std::string & id_parameter_name, unsigned int comp) const
+{
+  unsigned int id = getElementIDIndex(id_parameter_name, comp);
```
Comment (YaqiWang) [inline framework/src/interfaces/ElementIDInterface.C]: Intentionally not set up `SubProblem`, `Assembly`, `tid` in the constructor because I'd like the mesh generators able to use this interface for `getElementIDIndex`.

**Comment by YaqiWang** [comment]
Moved `ElementIDInterface` down from `UserObject` to `ElementUserObject` and checked again this PR. It is good to go to me. Could you please review it in your early convenience? Should not take much time.

**Comment by YaqiWang** [comment]
BTW, I did not add support for extra node IDs because it is not needed by our application. If someone want that, that can be done similarly as what I did in `ElementIDInterface`. Similarly for the extra datum recently added in libMesh.

**Comment by YaqiWang** [comment]
Just one thing pop up in my mind: should we let mesh generator add extra integers or we let MooseMesh do it? Currently in this PR, I have `extra_integers` as a parameter in `MooseMesh` and  `SubdomainBoundingBoxGenerator` uses it. We can just let `SubdomainBoundingBoxGenerator` add the integers. This is actually more desirable for my application.

**Comment by permcody** [comment]
I would definitely move the extra_integers to the class that needs them. As you pointed out, there's not really a good reason to put it on the mesh directly. That's a bit error prone.

**Comment by YaqiWang** [comment]
The only drawback I can see is that the extra integers will not be so obvious for a generator that is not declaring it. Extra integers are global to a mesh that to be generated. I am willing to make this change though because it will allow me to programmatically add extra integers in my specific generators.

**Comment by YaqiWang** [comment]
This is good to review again. I make `GeneratedMeshGenerator` to add the extra integers and let the `SubdomainBoundingBoxGenerator` to set the integer. Basically, the generators that do not actually generating a mesh should not have the extra integers. I suggest that in the future we create inheritance like

```
MeshingObject => MeshGenerator
                          \=>MeshModifier
```
and make `extra_integers` parameter available only in `MeshGenerator`.

**Comment by YaqiWang** [comment]
I will add another commit showing stitching two meshes with different set of integers will cause an error.

**Comment by YaqiWang** [comment]
Added a test for stitcher. Looks like it is doing the right thing in libMesh. We do not need to bother @roystgnr at all ;-)

**Comment by YaqiWang** [comment]
No, I was wrong. Stitching does not work. @roystgnr We will need you improve the libMesh interface to be able to resize the element integer vector and merge the element integers during stitching. I do not need this capability but with the new MOOSE mesh generation system. It will be nice to allow this. Or we can throw an error in the stitcher. The second commit in this PR shows the problem.

**Comment by roystgnr** [comment]
So what should happen in general when stitching two meshes?

If I start with ```foo1_idx = mesh1->add_node_integer("foo");``` and ```bar1_idx = mesh1->add_node_integer("bar");``` on one mesh, along with ```bar2_idx = mesh2->add_node_integer("bar");``` and ```baz2_idx = mesh2->add_node_integer("baz");``` for example, then what should be the state after ```mesh1->stitch_meshes(mesh2, b1, b2);```?  There's no way to make the differing ```mesh2``` integer indices line up on ```mesh1```, and no way to store two different integers with the same name, so IMHO the only thing to do here is let the "foo" and "bar" indices remain unchanged while appending "baz" to the end.  We could then add a "baz" entry (default 0) to all the original mesh1 nodes, and we could move the values for that entry as well as for "bar" on the newly-copied-to-mesh1 nodes so that they fall at the new indices rather than the mesh2 indices.  We'd then have to just trust that user code is smart enough to re-query the index for "bar" and not just try to use bar2_idx (or baz2_idx for that matter) on the stitched mesh.

Would that be acceptable?  Inadequate?  Overkill?

**Comment by YaqiWang** [comment]
I think it is acceptable. Typically query the index with an integer name happens after the mesh is constructed. Stitching is part of the mesh construction. User cannot use the index obtained before for accessing integers on mesh2.

**Comment by YaqiWang** [comment]
@permcody We need this to be merged soon, so that we can refactor the element ID support in Rattlesnake. We have a customer complaining restarting a calculation with element IDs (with the current implementation) does not work.

**Comment by YaqiWang** [comment]
Thanks!

**Comment by roystgnr** [comment]
So here's a question that came up while writing unit tests: when we stitch two meshes, and we have duplicate nodes on the stitched boundaries, which nodes' extra_integer data do we want to keep?  Both my intuition and the current code say that the nodes on ```this``` mesh should override the nodes on ```other_mesh``` but if we want to do something different then the time is now.

**Comment by roystgnr** [comment]
See if https://github.com/libMesh/libmesh/pull/2304 works for you?

**Comment by YaqiWang** [comment]
Although this PR is not using node ID, that is a very valid question to ask. I'd say we give user an option in the stitcher, that can default to keep the `this` mesh's node integers. Moose's `StitchedMeshGenerator` can pass a user's input along to the stitcher. Note we will have similar situation for extra node data. So the option can be used to control both.

**Comment by YaqiWang** [comment]
Thanks @roystgnr. I will report back soon.

**Comment by YaqiWang** [comment]
@roystgnr The test in my last commit works fine with https://github.com/libMesh/libmesh/pull/2304.

**Comment by roystgnr** [comment]
Thanks!  Let me just figure out why *my* last test isn't working and I'll get it in ASAP.

**Comment by YaqiWang** [comment]
I guess we are ready to make a final push on this PR.

**Comment by YaqiWang** [comment]
I will need to take a look on the failure with distributed mesh.

**Comment by roystgnr** [comment]
If you can replicate it, post a stack trace for me?

**Comment by YaqiWang** [comment]
Here is the back trace with 4 processors. All have the same.

```
  * frame #0: 0x000000010884e588 libmoose_test-dbg.0.dylib`libMesh::MeshBase::get_boundary_info() [inlined] std::__1::unique_ptr<libMesh::BoundaryInfo, std::__1::default_delete<libMesh::BoundaryInfo> >::operator*(this=0x0000000000000010) const at memory:2563
    frame #1: 0x000000010884e588 libmoose_test-dbg.0.dylib`libMesh::MeshBase::get_boundary_info(this=0x0000000000000000) at mesh_base.h:137
    frame #2: 0x0000000109e6c46e libmoose-dbg.0.dylib`StitchedMeshGenerator::generate(this=0x00007f8301c23808) at StitchedMeshGenerator.C:82
    frame #3: 0x000000010adda1e3 libmoose-dbg.0.dylib`MooseApp::executeMeshGenerators(this=0x00007f830001ca18) at MooseApp.C:1606
    frame #4: 0x00000001091b50ac libmoose-dbg.0.dylib`ExecuteMeshGenerators::act(this=0x00007f82ff309d18) at ExecuteMeshGenerators.C:28
    frame #5: 0x000000010914847a libmoose-dbg.0.dylib`Action::timedAct(this=0x00007f82ff309d18) at Action.C:90
    frame #6: 0x000000010915f724 libmoose-dbg.0.dylib`ActionWarehouse::executeActionsWithAction(this=0x00007f830001d438, task="execute_mesh_generators") at ActionWarehouse.C:364
    frame #7: 0x000000010915ea21 libmoose-dbg.0.dylib`ActionWarehouse::executeAllActions(this=0x00007f830001d438) at ActionWarehouse.C:338
    frame #8: 0x000000010adb88da libmoose-dbg.0.dylib`MooseApp::runInputFile(this=0x00007f830001ca18) at MooseApp.C:840
    frame #9: 0x000000010adba9f7 libmoose-dbg.0.dylib`MooseApp::run(this=0x00007f830001ca18) at MooseApp.C:988
    frame #10: 0x000000010846f9e3 moose_test-dbg`main(argc=10, argv=0x00007fff57790a70) at main.C:33
    frame #11: 0x00007fff8f73a235 libdyld.dylib`start + 1
```

**Comment by YaqiWang** [comment]
This seems from this line https://github.com/idaholab/moose/blob/next/framework/src/meshgenerators/StitchedMeshGenerator.C#L60. My question is when we generate mesh in `--distributed-mesh`, what exactly are we doing? All processor are generating the mesh and will be distributed properly later? I guess the mesh is in type of `DistributedMesh` but without elements distributed yet. This is completely fine with me. If so, we possibly need to cast to `MeshBase`.

**Comment by YaqiWang** [comment]
For now, I will restrict the test to replicated mesh. Making mesh generators work with distributed mesh will be a large task.

**Comment by YaqiWang** [comment]
Need this to be reviewed/merged. We have been waiting too long on this.

**Comment by permcody** [comment]
Alright then, Tagging @friedmud to take a look - We'll work on getting this in today provided there are not any more hangups.

File: framework/doc/content/syntax/Mesh/index.md [hunk 3]
```diff
@@ -165,4 +165,25 @@ recycled (because it might be a key to an important map), you should use unique_
...
+Extra integer IDs for all the elements of a mesh can be useful for handling complicated material assignment, performing specific calculations on groups of elements, etc.
+Often times, we do not want to use subdomain IDs for these tasks because otherwise too many subdomains could be needed, and in turn large penalty on run-time performance could be introduced.
+
+MooseMesh[MooseMesh.md] has a parameter `extra_integers` to allow users to introduce more integer IDs for elements each identified with a name in the parameter.
+When this parameter is specified, extra integers will be made available for all elements through `Assembly` in MOOSE objects such as kernels, aux kernels, materials, initial conditions, element user objects, etc.
+To retrieve the integer on an element, one needs to simply call
+```
+getElementID(integer_name_parameter, comp),
+```
+where `integer_name_parameter` is the name of the parameter in type of `std::vector<std::string>` of this object listing all integer names.
```
Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]:
```suggestion
within the initialization list of your constructor.  `integer_name_parameter` is the name of the parameter in type of `std::vector<std::string>` of this object listing all integer names.
```

File: framework/doc/content/syntax/Mesh/index.md [hunk 4]
```diff
@@ -165,4 +165,25 @@ recycled (because it might be a key to an important map), you should use unique_
...
+`comp` is the number of the integer names the returned $_id$ is for.
```
Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]:
```suggestion
`comp` is the index into the integer names if multiple are specified for this parameter.
```

File: framework/doc/content/syntax/Mesh/index.md [hunk 5]
```diff
@@ -165,4 +165,25 @@ recycled (because it might be a key to an important map), you should use unique_
...
+It is noticed that the returned value of this function call must be in type of `const dof_id_type &`, which is used to refer the value set by MOOSE in `Assembly`.
```
Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]:
```suggestion
It is noticed that the returned value of this function call must be in type of `const dof_id_type &`, which is used to refer the value set by MOOSE in `Assembly`.  The returned reference should be held in a class member variable for later use.
```

File: framework/doc/content/syntax/Mesh/index.md [hunk 6]
```diff
@@ -165,4 +165,25 @@ recycled (because it might be a key to an important map), you should use unique_
...
+Based on this ID, one can proceed with any particular operations, for example, choosing a different set of data for evaluating material properties.
+
+The IDs can be assigned to the mesh elements with mesh modifiers similarly as assigning subdomain IDs.
```
Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]:
```suggestion
IDs can be assigned to mesh elements with `MeshGenerators` in a similar way to assigning subdomain IDs.
```

File: framework/doc/content/syntax/Mesh/index.md [hunk 7]
```diff
@@ -165,4 +165,25 @@ recycled (because it might be a key to an important map), you should use unique_
...
+Currently we only extended SubdomainBoundingBoxGenerator[SubdomainBoundingBoxGenerator.md] for such assignment.
```
Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]: Let's remove all of this stuff about what has this capability and what might get it in the future.  We will forget to update this documentation as things change.

Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]:
```suggestion
```

File: framework/doc/content/syntax/Mesh/index.md [hunk 8]
```diff
@@ -165,4 +165,25 @@ recycled (because it might be a key to an important map), you should use unique_
...
+We possibly can enhance the extra integer assignment in the future for example, using an elemental variable in an Exodus file to assign integer IDs.
```
Comment (friedmud) [inline framework/doc/content/syntax/Mesh/index.md]:
```suggestion
```

File: framework/src/auxkernels/ElementIntegerAux.C [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+registerMooseObject("MooseApp", ElementIntegerAux);
+
+template <>
+InputParameters
+validParams<ElementIntegerAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addClassDescription("Creates a field showing the element integer.");
+  params.addParam<std::vector<std::string>>("integer_names", "Element integers to be retrieved");
```
Comment (friedmud) [inline framework/src/auxkernels/ElementIntegerAux.C]: It would be good to make `std::vector<std::string>` here just be `ExtraElementIDs`.  See how it's done in `MooseTypes`.

File: framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C [hunk 10]
```diff
@@ -55,19 +57,37 @@ SubdomainBoundingBoxGenerator::generate()
...
-    if (contains && _location == "INSIDE")
-      elem->subdomain_id() = _block_id;
-    else if (!contains && _location == "OUTSIDE")
-      elem->subdomain_id() = _block_id;
+    std::string integer_name = getParam<std::string>("integer_name");
+
+    if (!mesh->has_elem_integer(integer_name))
+      mooseError("Mesh does not have an element integer names as ", integer_name);
+
+    unsigned int id = mesh->get_elem_integer_index(integer_name);
```
Comment (friedmud) [inline framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C]: This should use the interface... right?

**Review by friedmud — CHANGES_REQUESTED** [review]
Getting there.

File: framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C (hunk 10 above)
Comment (YaqiWang) [reply framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C]: No with my second thought. I initially planed to use the interface, but the mesh generators get executed much early when the FEProblems, Assembly, stuff are not available. I think using the libMesh interface functions should be fine in mesh generators.

File: framework/src/auxkernels/ElementIntegerAux.C (hunk 9 above)
Comment (YaqiWang) [reply framework/src/auxkernels/ElementIntegerAux.C]: Ok. Is `ExtraElementIDNames` better?

File: framework/include/interfaces/ElementIDInterface.h [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+class MooseMesh;
+
+class ElementIDInterface
+{
+public:
+  ElementIDInterface(const MooseObject * moose_object);
+  virtual ~ElementIDInterface() {}
+
+  /**
+   * Gets index of an element integer with a parameter of the object this interface attachs
```
[the remark this answers is not in the document]
Comment (YaqiWang) [reply framework/include/interfaces/ElementIDInterface.h]: `attaches` means the object derived from this interface. Let me rephrase a little.

File: framework/src/auxkernels/ElementIntegerAux.C (hunk 9 above)
Comment (YaqiWang) [reply framework/src/auxkernels/ElementIntegerAux.C]: We have to have `std::vector<ExtraElementIDName>` here because MooseTypes.h only has `DerivativeStringClass`. But I think it is better than raw `std::string` already.

**Comment by YaqiWang** [comment]
Comments are address. Sorry, I should not do the rebase cause now you cannot easily see what I really changed in this new commit.

File: framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C (hunk 10 above)
Comment (friedmud) [reply framework/src/meshgenerators/SubdomainBoundingBoxGenerator.C]: Ok

**Review by friedmud — APPROVED** [review]

**Review by permcody — APPROVED** [review]

**Comment by YaqiWang** [comment]
😂
